// guilded.js/dist/managers/global/DocManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalDocManager = void 0;
const GlobalManager_1 = require("./GlobalManager");
class GlobalDocManager extends GlobalManager_1.GlobalManager {
    /** Create a doc. */
    create(channelId, options) {
        return this.client.rest.router.createDoc(channelId, options).then((data) => data.doc);
    }
    /** Get the docs from a channel. */
    fetchMany(channelId) {
        return this.client.rest.router.getDocs(channelId).then((data) => data.docs);
    }
    /** Get a doc from a channel. */
    fetch(channelId, docId) {
        return this.client.rest.router.getDoc(channelId, docId).then((data) => data.doc);
    }
    /** Update a doc */
    update(channelId, docId, options) {
        return this.client.rest.router.updateDoc(channelId, docId, options).then((data) => data.doc);
    }
    /** Delete a doc from a channel. */
    delete(channelId, docId) {
        return this.client.rest.router.deleteDoc(channelId, docId).then(() => void 0);
    }
}
exports.GlobalDocManager = GlobalDocManager;
