// guilded.js/dist/managers/global/ListManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalListItemManager = void 0;
const GlobalManager_1 = require("./GlobalManager");
class GlobalListItemManager extends GlobalManager_1.GlobalManager {
    /** Create a list item. */
    create(channelId, options) {
        return this.client.rest.router.createListItem(channelId, options).then((data) => data.listItem);
    }
    /** Get list items */
    fetchMany(channelId) {
        return this.client.rest.router.getListItems(channelId).then((data) => data.listItems);
    }
    /** Get list item */
    fetch(channelId, itemId) {
        return this.client.rest.router.getListItem(channelId, itemId).then((data) => data.listItem);
    }
    /** Update list item */
    update(channelId, itemId, options) {
        return this.client.rest.router.updateListItem(channelId, itemId, options).then((data) => data.listItem);
    }
    /** Delete list item */
    delete(channelId, itemId) {
        return this.client.rest.router.deleteListItem(channelId, itemId).then(() => void 0);
    }
    /** Complete list item */
    complete(channelId, itemId) {
        return this.client.rest.router.completeListItem(channelId, itemId).then(() => void 0);
    }
    /** Uncomplete list item */
    uncomplete(channelId, itemId) {
        return this.client.rest.router.uncompleteListItem(channelId, itemId).then(() => void 0);
    }
}
exports.GlobalListItemManager = GlobalListItemManager;
