// guilded.js/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ROUTES = exports.constants = void 0;
exports.constants = {
    clientEvents: {
        MESSAGE_CREATED: "messageCreated",
        MESSAGE_UPDATED: "messageUpdated",
        MESSAGE_DELETED: "messageDeleted",
        MEMBER_UPDATED: "memberUpdated",
        ROLES_UPATED: "rolesUpdated",
        MEMBER_JOINED: "memberJoined",
        MEMBER_REMOVED: "memberRemoved",
        MEMBER_BANNED: "memberBanned",
        MEMBER_UNBANNED: "memberUnbanned",
        WEBHOOK_CREATED: "webhookCreated",
        WEBHOOK_UPDATED: "webhookUpdated",
        LIST_ITEM_COMPLETED: "listItemCompleted",
        LIST_ITEM_UNCOMPLETED: "listItemUncompleted",
        LIST_ITEM_CREATED: "listItemCreated",
        LIST_ITEM_UPDATED: "listItemUpdated",
        LIST_ITEM_DELETED: "listItemDeleted",
        CHANNEL_CREATED: "channelCreated",
        CHANNEL_UPDATED: "channelUpdated",
        CHANNEL_DELETED: "channelDeleted",
        DOC_CREATED: "docCreated",
        DOC_UPDATED: "docUpdated",
        DOC_DELETED: "docDeleted",
        SERVER_CREATED: "serverCreated",
    },
};
exports.ROUTES = {
    AWS_CDN: "https://s3-us-west-2.amazonaws.com/www.guilded.gg/",
    BASE_DOMAIN: "www.guilded.gg",
    API_DOMAIN: "api.guilded.gg",
    CDN: "img.guildedcdn.com",
    IMAGE_CDN_DOMAIN: "img.guildedcdn.com",
    MEDIA_DOMAIN: "media.guilded.gg",
};
