// @guildedjs/gil/dist/utils/walk.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.walk = void 0;
const colorette_1 = require("colorette");
const promises_1 = require("fs/promises");
const path_1 = __importDefault(require("path"));
/** Walks through a directory allowing you to process it with a for await loop */
function walk(dir) {
    return __asyncGenerator(this, arguments, function* walk_1() {
        var e_1, _a;
        const folder = yield __await((0, promises_1.opendir)(dir).catch((error) => {
            if (error.message.startsWith("ENOENT: no such file or directory")) {
                console.log((0, colorette_1.yellowBright)(`[WARN] Missing folder: ${dir.substring(dir.lastIndexOf("/"))}. To make this warning go away, simply create the folder in your src folder.`));
            }
            else
                console.log(error);
        }));
        if (!folder)
            return yield __await(void 0);
        const directories = [];
        try {
            for (var folder_1 = __asyncValues(folder), folder_1_1; folder_1_1 = yield __await(folder_1.next()), !folder_1_1.done;) {
                const d = folder_1_1.value;
                const entry = path_1.default.join(dir, d.name);
                // if (d.isDirectory()) yield* walk(entry);
                if (d.isDirectory())
                    directories.push(entry);
                if (![".js", ".ts"].some((suffix) => entry.endsWith(suffix)))
                    continue;
                else if (d.isFile())
                    yield yield __await([d.name, require(entry)]);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (folder_1_1 && !folder_1_1.done && (_a = folder_1.return)) yield __await(_a.call(folder_1));
            }
            finally { if (e_1) throw e_1.error; }
        }
        // LOAD DIRECTORIES AFTER FILES TO ALLOW SUBCOMMANDS TO HAVE COMMAND FILES MADE FIRST!
        for (const directory of directories)
            yield __await(yield* __asyncDelegator(__asyncValues(walk(directory))));
    });
}
exports.walk = walk;
exports.default = walk;
