// guilded.js/dist/structures/channels/Channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = void 0;
const Base_1 = require("../Base");
class Channel extends Base_1.Base {
    constructor(client, data) {
        super(client, data);
        this.serverId = data.serverId;
        this.type = channelTypeToEnumMap[data.type];
        this.createdAt = new Date(data.createdAt);
        this.createdBy = data.createdBy;
        this.groupId = data.groupId;
        this._update(data);
    }
    _update(data) {
        var _a, _b, _c, _d, _e;
        if ("name" in data && typeof data.name !== "undefined") {
            this.name = data.name;
        }
        if ("topic" in data) {
            this.topic = (_a = data.topic) !== null && _a !== void 0 ? _a : null;
        }
        if ("updatedAt" in data && typeof data.updatedAt !== "undefined") {
            this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        }
        if ("parentId" in data && typeof data.updatedAt !== "undefined") {
            this.parentId = (_b = data.parentId) !== null && _b !== void 0 ? _b : null;
        }
        if ("categoryId" in data && typeof data.categoryId !== "undefined") {
            this.categoryId = (_c = data.categoryId) !== null && _c !== void 0 ? _c : null;
        }
        if ("isPublic" in data && typeof data.isPublic !== "undefined") {
            this.isPublic = (_d = data.isPublic) !== null && _d !== void 0 ? _d : false;
        }
        if ("archivedBy" in data && typeof data.archivedBy !== "undefined") {
            this.archivedBy = (_e = data.archivedBy) !== null && _e !== void 0 ? _e : null;
        }
        if ("archivedAt" in data && typeof data.archivedAt !== "undefined") {
            this.archivedAt = data.archivedAt ? new Date(data.archivedAt) : null;
        }
        return this;
    }
    /** Get a list of the latest 100 messages from a channel. */
    fetchMessages(options) {
        return this.client.messages.fetchMany(this.id, options !== null && options !== void 0 ? options : {});
    }
    /** Get details for a specific chat message from a chat channel. */
    fetchMessage(messageId) {
        return this.client.messages.fetch(this.id, messageId);
    }
    /** Update this channel. */
    update(options) {
        return this.client.channels.update(this.id, options);
    }
    /** Delete this channel. */
    delete() {
        return this.client.channels.delete(this.id);
    }
    /** Send a chat message in the channel. */
    send(content) {
        return this.client.messages.send(this.id, content);
    }
}
exports.Channel = Channel;
var ChannelType;
(function (ChannelType) {
    ChannelType[ChannelType["Announcements"] = 0] = "Announcements";
    ChannelType[ChannelType["Chat"] = 1] = "Chat";
    ChannelType[ChannelType["Calendar"] = 2] = "Calendar";
    ChannelType[ChannelType["Forums"] = 3] = "Forums";
    ChannelType[ChannelType["Media"] = 4] = "Media";
    ChannelType[ChannelType["Docs"] = 5] = "Docs";
    ChannelType[ChannelType["Voice"] = 6] = "Voice";
    ChannelType[ChannelType["List"] = 7] = "List";
    ChannelType[ChannelType["Scheduling"] = 8] = "Scheduling";
    ChannelType[ChannelType["Stream"] = 9] = "Stream";
})(ChannelType || (ChannelType = {}));
const channelTypeToEnumMap = {
    announcements: ChannelType.Announcements,
    chat: ChannelType.Chat,
    calendar: ChannelType.Calendar,
    forums: ChannelType.Forums,
    media: ChannelType.Media,
    docs: ChannelType.Docs,
    voice: ChannelType.Voice,
    list: ChannelType.List,
    scheduling: ChannelType.Scheduling,
    stream: ChannelType.Stream,
};
