// guilded.js/dist/structures/channels/DocChannel.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocChannel = void 0;
const collection_1 = __importDefault(require("@discordjs/collection"));
const Channel_1 = require("./Channel");
class DocChannel extends Channel_1.Channel {
    constructor() {
        super(...arguments);
        /** The docs in this channel. */
        this.docs = new collection_1.default();
    }
    /** Create a doc. */
    createDoc(options) {
        return this.client.docs.create(this.id, options);
    }
    /** Get the docs from this channel. */
    getDocs() {
        return this.client.docs.fetchMany(this.id);
    }
    /** Get a doc from this channel. */
    getDoc(docId) {
        return this.client.docs.fetch(this.id, docId);
    }
    /** Update a doc in this channel. */
    updateDoc(docId, options) {
        return this.client.docs.update(this.id, docId, options);
    }
    /** Delete a doc from this channel. */
    deleteDoc(docId) {
        return this.client.docs.delete(this.id, docId);
    }
}
exports.DocChannel = DocChannel;
