// guilded.js/dist/structures/channels/ListChannel.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListChannel = void 0;
const collection_1 = __importDefault(require("@discordjs/collection"));
const Channel_1 = require("./Channel");
class ListChannel extends Channel_1.Channel {
    constructor() {
        super(...arguments);
        /** The items in this channel. */
        this.items = new collection_1.default();
    }
    /** Create a list item. */
    createItem(message, note) {
        return this.client.lists.create(this.id, { message, note });
    }
    getItem(itemId) {
        return this.client.lists.fetch(this.id, itemId).then((data) => {
            this.items.set(data.id, data);
            return data;
        });
    }
    getItems() {
        return this.client.lists.fetchMany(this.id).then((data) => {
            for (const item of data) {
                this.items.set(item.id, item);
            }
            return data;
        });
    }
    completeItem(itemId) {
        return this.client.lists.complete(this.id, itemId).then(() => void 0);
    }
    uncompletedItem(itemId) {
        return this.client.lists.uncomplete(this.id, itemId).then(() => void 0);
    }
}
exports.ListChannel = ListChannel;
