// guilded.js/dist/structures/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserType = exports.ClientUser = exports.User = void 0;
const Base_1 = require("./Base");
class User extends Base_1.Base {
    constructor(client, data) {
        super(client, data);
        /** The avatar image associated with this user */
        this.avatar = null;
        /** The banner image associated with this user */
        this.banner = null;
        this.name = data.name;
        this.createdAt = new Date(data.createdAt);
        this.type = data.type === "bot" ? UserType.Bot : UserType.User;
        this._update(data);
    }
    _update(data) {
        var _a, _b;
        if ("avatar" in data) {
            this.avatar = (_a = data.avatar) !== null && _a !== void 0 ? _a : null;
        }
        if ("banner" in data) {
            this.banner = (_b = data.banner) !== null && _b !== void 0 ? _b : null;
        }
        return this;
    }
}
exports.User = User;
class ClientUser extends User {
    constructor(client, data) {
        super(client, Object.assign(Object.assign({}, data), { type: "bot" }));
        this.createdBy = data.createdBy;
        this.botId = data.botId;
    }
}
exports.ClientUser = ClientUser;
var UserType;
(function (UserType) {
    UserType[UserType["Bot"] = 0] = "Bot";
    UserType[UserType["User"] = 1] = "User";
})(UserType = exports.UserType || (exports.UserType = {}));
