// @guildedjs/guilded-api-typings/dist/v1/ws/Events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketEvents = exports.WSOpCodes = void 0;
var WSOpCodes;
(function (WSOpCodes) {
    WSOpCodes[WSOpCodes["SUCCESS"] = 0] = "SUCCESS";
    WSOpCodes[WSOpCodes["WELCOME"] = 1] = "WELCOME";
    WSOpCodes[WSOpCodes["RESUME"] = 2] = "RESUME";
    WSOpCodes[WSOpCodes["ERROR"] = 8] = "ERROR";
    WSOpCodes[WSOpCodes["PING"] = 9] = "PING";
    WSOpCodes[WSOpCodes["PONG"] = 10] = "PONG";
})(WSOpCodes = exports.WSOpCodes || (exports.WSOpCodes = {}));
exports.WebSocketEvents = {
    ChatMessageCreated: "ChatMessageCreated",
    ChatMessageUpdated: "ChatMessageUpdated",
    ChatMessageDeleted: "ChatMessageDeleted",
    TeamMemberJoined: "TeamMemberJoined",
    TeamMemberRemoved: "TeamMemberRemoved",
    TeamMemberUpdated: "TeamMemberUpdated",
    TeamMemberBanned: "TeamMemberBanned",
    TeamMemberUnbanned: "TeamMemberUnbanned",
    // This is intentional. Legacy change on Guilded's end.
    teamRolesUpdated: "teamRolesUpdated",
    TeamWebhookCreated: "TeamWebhookCreated",
    TeamWebhookUpdated: "TeamWebhookUpdated",
    ListItemCompleted: "ListItemCompleted",
    ListItemUncompleted: "ListItemUncompleted",
    ListItemCreated: "ListItemCreated",
    ListItemUpdated: "ListItemUpdated",
    ListItemDeleted: "ListItemDeleted",
    TeamChannelCreated: "TeamChannelCreated",
    TeamChannelUpdated: "TeamChannelUpdated",
    TeamChannelDeleted: "TeamChannelDeleted",
    DocCreated: "DocCreated",
    DocUpdated: "DocUpdated",
    DocDeleted: "DocDeleted",
};
